### PR TITLE
Poison CUmodule so that we notice memory corruption

### PR DESCRIPTION
**Pull request by tfogal**

### CODE REVIEW

**Comment by tfogal** [comment]
!build

**Comment by zasdfgbnm** [comment]
Sorry, I don't understand this PR. Why we would double-unload module, and why this PR avoid it?

File: csrc/executor_utils.cpp [hunk 1]
```diff
@@ -1151,6 +1151,9 @@ CompiledKernel::~CompiledKernel() {
   if (module != nullptr) {
     NVFUSER_CUDA_SAFE_CALL(cuModuleUnload(module));
   }
+#ifndef NDEBUG
+  module = nullptr;
+#endif
```
Comment (wujingyue) [inline csrc/executor_utils.cpp:1156]:
```suggestion
  module = nullptr;
```

I'd do this

File: csrc/executor_utils.cpp [hunk 2]
```diff
@@ -1151,6 +1151,9 @@ CompiledKernel::~CompiledKernel() {
   if (module != nullptr) {
     NVFUSER_CUDA_SAFE_CALL(cuModuleUnload(module));
   }
+#ifndef NDEBUG
```
Comment (wujingyue) [inline csrc/executor_utils.cpp:1154]: if (module != nullptr) {
  unload
  module = nullptr;
}

Also, I'd remove NDEBUG unless you find it indeed expensive to set null.

**Review by wujingyue — APPROVED** [review]

**Comment by tfogal** [comment]
> > > Why we would double-unload module, and why this PR avoid it?
> > 
> > The situation was, I believe, a memory corruption situation.
> > This avoids it because of the `if` statement---see right above the added lines.
> 
> This is in the destructor of `CompiledKernel`. IIUC, destructor is tearing down the life time of the object. So it will make no difference whether or not we set the pointer to nullptr. Are we calling the destructor multiple times? This doesn't sound correct. Or is this just some trick to help debugging? https://stackoverflow.com/a/4666535

Yeah, I like the idea of setting it to a known-bad value; I've seen value from doing that in the past. Will update.

**Comment by tfogal** [comment]
!build

**Comment by wujingyue** [comment]
> Yeah, I like the idea of setting it to a known-bad value; I've seen value from doing that in the past. Will update.

Make sure you update the PR title. It's accurate any more after your latest change. :shipit:

**Comment by tfogal** [comment]
Hey @jjsjann123, is it possible this is coming from somewhere else?

with:

> commit eb42bdb7624385d53237d0737a10b334ee2e79d8 (HEAD, jie/to_memory_format)
> Author: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
> Date:   Wed Apr 10 23:08:55 2024 +0000
> 
>     [pre-commit.ci] auto fixes from pre-commit.com hooks
>     
>     for more information, see https://pre-commit.ci

 I am still getting:
```
[rank0]:     super().optimizer_step(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py", line 1303, in optimizer_step
[rank0]:     optimizer.step(closure=optimizer_closure)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/optimizer.py", line 152, in step
[rank0]:     step_output = self._strategy.optimizer_step(self._optimizer, closure, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/ddp.py", line 270, in optimizer_step
[rank0]:     optimizer_output = super().optimizer_step(optimizer, closure, model, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 239, in optimizer_step
[rank0]:     return self.precision_plugin.optimizer_step(optimizer, model=model, closure=closure, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/plugins/precision/amp.py", line 80, in optimizer_step
[rank0]:     closure_result = closure()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/optimization/automatic.py", line 144, in __call
__
[rank0]:     self._result = self.closure(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/optimization/automatic.py", line 129, in closur
e
[rank0]:     step_output = self._step_fn()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/optimization/automatic.py", line 318, in _train
ing_step     
[rank0]:     training_step_output = call._call_strategy_hook(trainer, "training_step", *kwargs.values())
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py", line 309, in _call_strategy_hook
[rank0]:     output = fn(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 390, in training_step
[rank0]:     return self._forward_redirection(self.model, self.lightning_module, "training_step", *args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 642, in __call__
[rank0]:     wrapper_output = wrapper_module(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 635, in wrapped_forward
[rank0]:     out = method(*_args, **_kwargs)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/utils/model_utils.py", line 399, in wrap_training_step
[rank0]:     output_dict = wrapped(*args, **kwargs)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py", line 393, in training
_step
[rank0]:     loss_mean, loss_dict = self.fwd_bwd_step(dataloader_iter, False)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py", line 346, in fwd_bwd_
step
[rank0]:     losses_reduced_per_micro_batch = fwd_bwd_function(
[rank0]:   File "/home/tfogal/dev/mlm/megatron/core/pipeline_parallel/schedules.py", line 374, in forward_backward_no_pipelining
[rank0]:     output_tensor = forward_step(
[rank0]:   File "/home/tfogal/dev/mlm/megatron/core/pipeline_parallel/schedules.py", line 192, in forward_step
[rank0]:     output_tensor, loss_func = forward_step_func(data_iterator, model)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py", line 254, in fwd_outp
ut_and_loss_func
[rank0]:     loss, loss_dict = model(*batch)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 194, in forward
[rank0]:     res = self._forward_fn(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 629, in fn_
[rank0]:     cache_entry, inps, pro_to_epi = get_computation_and_inputs(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 262, in cache_info_wrapper
[rank0]:     res = fn(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 504, in get_computation_and_inputs
[rank0]:     prologue_trc, computation_trc, *maybe_epilogue = interpreter(
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 175, in _general_frontend
[rank0]:     return thunder_general_jit(fn, args, kwargs, sharp_edges=sharp_edges)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/jit_ext.py", line 1440, in thunder_general_jit
[rank0]:     result = jfn(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6684, in fn_
[rank0]:     raise e
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6647, in fn_2
[rank0]:     return fn(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6046, in _impl
[rank0]:     return fn.__func__(fn.__self__, *args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6046, in _impl
[rank0]:     return fn.__func__(fn.__self__, *args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6046, in _impl
[rank0]:     return fn.__func__(fn.__self__, *args, **kwargs)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py", line 135, in forward
[rank0]:     x_start = x_start.to(memory_format=torch.channels_last)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6158, in partial_call_impl
[rank0]:     return partial_function.func(*(partial_function.args + args), **(partial_function.keywords | kwargs))
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 1258, in wrapping_wrapper
[rank0]:     res = ufn(*uargs, **ukwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/symbol.py", line 250, in __call__
[rank0]:     result = self.meta(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/langctxs.py", line 124, in _fn
[rank0]:     result = fn(*args, **kwargs)
[rank0]: TypeError: to() got an unexpected keyword argument 'memory_format'
```

from the NeMo model. Did I apply it wrong somehow?

**Pull request merged**